[NNC] Add logical Operators '&&' and '||'

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56947 [NNC] Add logical Operators '&&' and '||'**

Differential Revision: [D28007342](https://our.internmc.facebook.com/intern/diff/D28007342)

huiguoo committed:
[NNC] Add logical Operators '&&' and '||'

[ghstack-poisoned]

huiguoo commented on torch/csrc/jit/tensorexpr/expr.cpp:
```diff
@@ -50,6 +50,36 @@ ExprHandle ExprHandle::operator<=(const ExprHandle& other) const {
+  if (!compare->is_ret_default()) {
+    throw std::runtime_error("invalid && operator");
+  }
+  CompareSelect* ret = new CompareSelect(
+      compare->lhs(),
+      compare->rhs(),
+      other.node(),
+      new IntImm(0),
+      compare->compare_select_op());
```
Thanks for the pointer @ZolotukhinM ! We probably need to add one comment for `CompareSelect` stating that the true expressions and false expressions are evaluated simultaneously. I was not aware of it and tried to avoid introducing the extra NOP needed in `IfThenElse`. 
Will switch to `IfThenElse`.

huiguoo committed:
Update on "[NNC] Add logical Operators '&&' and '||'"


Differential Revision: [D28007342](https://our.internmc.facebook.com/intern/diff/D28007342)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/expr.cpp:
```diff
@@ -50,6 +50,20 @@ ExprHandle ExprHandle::operator<=(const ExprHandle& other) const {
+  if (!this->node()->dtype().is_integral()) {
+    throw unsupported_dtype();
+  }
+  return IfThenElse::make(*this, IntImm::make(1), other);
```
The immediate must have the same dtype as `other`, otherwise IR will be malformed. For instance, if `other` is `bool`, then we need to use `BoolImm::make(1)` instead of `IntImm::make(1)`. Also I think it would be worth adding a test for exercising different dtypes as well.

huiguoo committed:
Update on "[NNC] Add logical Operators '&&' and '||'"


Differential Revision: [D28007342](https://our.internmc.facebook.com/intern/diff/D28007342)

[ghstack-poisoned]

huiguoo committed:
Update on "[NNC] Add logical Operators '&&' and '||'"


Differential Revision: [D28007342](https://our.internmc.facebook.com/intern/diff/D28007342)

[ghstack-poisoned]

huiguoo committed:
Update on "[NNC] Add logical Operators '&&' and '||'"


Differential Revision: [D28007342](https://our.internmc.facebook.com/intern/diff/D28007342)

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
LGTM!

huiguoo committed:
Update on "[NNC] Add logical Operators '&&' and '||'"


Differential Revision: [D28007342](https://our.internmc.facebook.com/intern/diff/D28007342)

[ghstack-poisoned]
